### PR TITLE
Add support for expanding dimensions using `None`

### PR DESCRIPTION
**Pull request by kgryte**
This PR

-   resolves https://github.com/data-apis/array-api/issues/360 by adding support for expanding dimensions via `None`. The functional equivalent of the proposed changes is repeated invocation of `expand_dims`. Using `None` in a selection tuple is more convenient when needing to insert multiple size `1` dimensions.
-   as per [comment](https://github.com/data-apis/array-api/issues/360#issuecomment-1075608356), using `None` in selection tuples enjoys broad support across array libraries.

### CODE REVIEW

**Review by rgommers — COMMENTED** [review]
I think adding `None` makes sense. My main question is whether it's okay to combine it with other indexing semantics. E.g,:
```python
>>> import torch
>>> t = torch.arange(5)
>>> t[None, :3, None]
tensor([[[0],
         [1],
         [2]]])
```
That probably needs testing and specifying.

**Comment by asmeurer** [comment]
> I think adding None makes sense. My main question is whether it's okay to combine it with other indexing semantics. E.g,:

There shouldn't be any issues with this I would imagine. The text from the [NumPy docs](https://numpy.org/doc/stable/user/basics.indexing.html) seems fine, although it could perhaps be more precise:

> Each [newaxis](https://numpy.org/doc/stable/reference/constants.html#numpy.newaxis) object in the selection tuple serves to expand the dimensions of the resulting selection by one unit-length dimension. The added dimension is the position of the [newaxis](https://numpy.org/doc/stable/reference/constants.html#numpy.newaxis) object in the selection tuple. [newaxis](https://numpy.org/doc/stable/reference/constants.html#numpy.newaxis) is an alias for None, and None can be used in place of this with the same result.

This should probably not apply to boolean indices where we should keep the "only index" rule.

File: spec/API_specification/indexing.rst [hunk 1]
```diff
@@ -158,6 +158,11 @@ Multi-dimensional arrays must extend the concept of single-axis indexing to mult
 
     To perform flat indexing, use ``reshape(x, (-1,))[integer]``.
 
+- Each ``None`` in the selection tuple must expand the dimensions of the resulting selection by one dimension of size ``1``. The position of the added dimension must be the same as the position of ``None`` in the selection tuple.
```
Comment (asmeurer) [inline spec/API_specification/indexing.rst:161]: Do we need to clarify that None is not a "single-axis" index for the purposes of the other rule (like the one below this one)?

Comment (kgryte) [reply spec/API_specification/indexing.rst]: I've updated the guidance for the `IndexError` rule. This should now be resolved.

**Comment by kgryte** [comment]
I've added an explicit note that the use of `None` in combination with boolean array indexing is not allowed. This should address @asmeurer's comment above.

**Comment by kgryte** [comment]
Re: combining with other indexing semantics. As @asmeurer mentions, this should not be an issue, given that the use of `None` is syntactic sugar for first extracting a selection and then applying `expand_dims`. Meaning, an implementation could filter a selection tuple for non-`None` values, compute the selection, and then apply `expand_dims` to the result. As such, there should be no conflict with using `None` in conjunction with other single-axis indexing expressions.

**Comment by asmeurer** [comment]
This looks good. Were we going to add `newaxis`?

**Comment by kgryte** [comment]
As this PR has been open for 3 weeks, has received approvals here, and no objections were raised during consortium meetings, will go ahead and merge. Any changes can be addressed in follow-up pull requests.
